hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationLog.java

```diff
@@ -0,0 +1,196 @@
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.util.JsonUtil.getArray;
+import static com.hazelcast.util.JsonUtil.getInt;
+import static com.hazelcast.util.JsonUtil.getLong;
+import static com.hazelcast.util.JsonUtil.getString;
+
+public final class SlowOperationLog implements JsonSerializable {
+    private static final int SHORT_STACKTRACE_LENGTH = 200;
+
+    private static final ConstructorFunction<Integer, Invocation> INVOCATION_CONSTRUCTOR_FUNCTION
+            = new ConstructorFunction<Integer, Invocation>() {
+        @Override
+        public Invocation createNew(Integer arg) {
+            return new Invocation();
+        }
+    };
+
+    private final ConcurrentHashMap<Integer, Invocation> invocations = new ConcurrentHashMap<Integer, Invocation>();
+
+    private String operation;
+    private String stackTrace;
+    private int totalInvocations;
+
+    private transient String shortStackTrace;
+
+    public SlowOperationLog(String stackTrace, Object operation) {
+        this.operation = operation.toString();
+        this.stackTrace = stackTrace;
+        this.totalInvocations = 0;
+    }
+
+    String getOperation() {
+        return operation;
+    }
+
+    String getStackTrace() {
+        return stackTrace;
+    }
+
+    String getShortStackTrace() {
+        if (shortStackTrace == null) {
+            shortStackTrace = stackTrace.substring(0, stackTrace.indexOf('\n', SHORT_STACKTRACE_LENGTH)) + "\n(...)";
+        }
+        return shortStackTrace;
+    }
+
+    int getTotalInvocations() {
+        return totalInvocations;
+    }
+
+    Collection<Invocation> getInvocations() {
+        return invocations.values();
+    }
+
+    Invocation getOrCreateInvocation(int operationHashCode, long lastDurationNanos, long nowNanos, long nowMillis) {
+        totalInvocations++;
+
+        Invocation invocation = ConcurrencyUtil.getOrPutIfAbsent(invocations, operationHashCode, INVOCATION_CONSTRUCTOR_FUNCTION);
+        invocation.id = operationHashCode;
+        invocation.duration = TimeUnit.NANOSECONDS.toMillis(lastDurationNanos);
+        invocation.startedAt = nowMillis - invocation.duration;
+        invocation.lastAccess = nowNanos;
+
+        return invocation;
+    }
+
+    void purgeInvocations(long now, long slowOperationLogLifetimeNanos) {
+        for (Map.Entry<Integer, Invocation> invocationEntry : invocations.entrySet()) {
+            if (now - invocationEntry.getValue().lastAccess > slowOperationLogLifetimeNanos) {
+                invocations.remove(invocationEntry.getKey());
+            }
+        }
+    }
+
+    boolean isEmpty() {
+        return invocations.isEmpty();
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append("SlowOperationLog{operation='").append(operation).append("', stackTrace='").append(stackTrace)
+          .append("', totalInvocations=").append(totalInvocations).append(", invocations='");
+
```

Comment:
What do you think of the following layout:

```
sb.append("SlowOperationLog{operation='").append(operation)
	.append("', stackTrace='").append(stackTrace)   
	.append("', totalInvocations=").append(totalInvocations)
	.append(", invocations='");
```

This is a lot easier to read because you can now see what belongs to each other. Long chaining of .append(..) are imho a PITA.